agb/src/display/affine.rs

```diff
@@ -0,0 +1,177 @@
+    ops::{Mul, MulAssign},
+};
+
+use agb_fixnum::{Num, Vector2D};
+
+type AffineMatrixElement = Num<i32, 8>;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct AffineMatrix {
+    a: AffineMatrixElement,
+    b: AffineMatrixElement,
+    c: AffineMatrixElement,
+    d: AffineMatrixElement,
+    x: AffineMatrixElement,
+    y: AffineMatrixElement,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct OverflowError(pub(crate) ());
+
+impl AffineMatrix {
+    #[must_use]
+    pub fn identity() -> Self {
+        AffineMatrix {
+            a: 1.into(),
+            b: 0.into(),
+            c: 0.into(),
+            d: 1.into(),
+            x: 0.into(),
+            y: 0.into(),
+        }
+    }
+
+    #[must_use]
+    pub fn from_rotation(angle: Num<i32, 16>) -> Self {
+        let cos = angle.cos().change_base();
+        let sin = angle.sin().change_base();
+
+        AffineMatrix {
+            a: cos,
+            b: sin,
+            c: -sin,
+            d: cos,
+            x: 0.into(),
+            y: 0.into(),
+        }
+    }
+
+    // Identity for rotation / scale / skew
+    #[must_use]
+    pub fn from_position(position: Vector2D<Num<i32, 8>>) -> Self {
+        AffineMatrix {
+            a: 1.into(),
+            b: 0.into(),
+            c: 0.into(),
+            d: 1.into(),
+            x: position.x,
+            y: position.y,
+        }
+    }
+
+    #[must_use]
+    pub fn position(&self) -> Vector2D<Num<i32, 8>> {
+        (self.x, self.y).into()
+    }
+
+    #[must_use]
+    pub fn try_to_background(&self) -> Option<AffineMatrixBackground> {
+        Some(AffineMatrixBackground {
+            a: self.a.to_raw().try_into().ok()?,
+            b: self.a.to_raw().try_into().ok()?,
+            c: self.a.to_raw().try_into().ok()?,
+            d: self.a.to_raw().try_into().ok()?,
+            x: self.a.to_raw(),
+            y: self.a.to_raw(),
+        })
+    }
+
+    #[must_use]
+    pub fn to_background_wrapping(&self) -> AffineMatrixBackground {
+        AffineMatrixBackground {
+            a: self.a.to_raw() as i16,
+            b: self.a.to_raw() as i16,
+            c: self.a.to_raw() as i16,
+            d: self.a.to_raw() as i16,
+            x: self.a.to_raw(),
+            y: self.a.to_raw(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(C, packed(4))]
+pub struct AffineMatrixBackground {
+    // Internally these can be thought of as Num<i16, 8>
+    a: i16,
+    b: i16,
+    c: i16,
+    d: i16,
+    // These are Num<i32, 8>
+    x: i32,
+    y: i32,
+}
+
+impl AffineMatrixBackground {
+    #[must_use]
+    pub fn to_affine_matrix(&self) -> AffineMatrix {
+        AffineMatrix {
+            a: Num::from_raw(self.a.into()),
+            b: Num::from_raw(self.b.into()),
+            c: Num::from_raw(self.c.into()),
+            d: Num::from_raw(self.d.into()),
+            x: Num::from_raw(self.x),
+            y: Num::from_raw(self.y),
+        }
+    }
+}
```

Comment:
should this just be the `impl From` rather than having this method?

Reply:
I included it because impl from can be hard to discover whereas this function is more obvious on docs.rs and hints from ide.